core/src/main/scala/kafka/log/Log.scala

```diff
@@ -2096,10 +2100,14 @@ class Log(@volatile private var _dir: File,
             val deletable = logSegments.filter(segment => segment.baseOffset > targetOffset)
             removeAndDeleteSegments(deletable, asyncDelete = true, LogTruncation)
             activeSegment.truncateTo(targetOffset)
-            updateLogEndOffset(targetOffset)
-            updateLogStartOffset(math.min(targetOffset, this.logStartOffset))
             leaderEpochCache.foreach(_.truncateFromEnd(targetOffset))
+
+            // Load producer state prior to adjusting log start and end offsets
+            // in case the first unstable offset is invalidated by the truncation.
             loadProducerState(targetOffset, reloadFromCleanShutdown = false)
```

Comment:
In the case of requiring truncation below the high watermark, ```maybeIncrementFirstUnstableOffset``` gets called two times (by ```loadProducerState``` and ```updateLogEndOffset```). Is first call unnecessary?

Reply:
Thanks, I think this is addressed in the latest commit.